oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/grpc/GRPCCallback.java

```diff
@@ -0,0 +1,102 @@
+
+import io.grpc.stub.StreamObserver;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.skywalking.oap.server.core.alarm.AlarmCallback;
+import org.apache.skywalking.oap.server.core.alarm.AlarmMessage;
+import org.apache.skywalking.oap.server.core.alarm.grpc.AlarmServiceGrpc;
+import org.apache.skywalking.oap.server.core.alarm.grpc.Response;
+import org.apache.skywalking.oap.server.core.alarm.provider.AlarmRulesWatcher;
+import org.apache.skywalking.oap.server.library.client.grpc.GRPCClient;
+
+/**
+ * Use SkyWalking alarm grpc API call a remote methods.
+ *
+ * @author jian.tan
+ */
+@Slf4j
+public class GRPCCallback implements AlarmCallback {
+
+    GRPCAlarmSetting alarmSetting;
+
+    private AlarmServiceGrpc.AlarmServiceStub alarmServiceStub;
+
+    public GRPCCallback(AlarmRulesWatcher alarmRulesWatcher) {
+        alarmSetting = alarmRulesWatcher.getGrpchookSetting();
+        if (alarmSetting != null) {
+            GRPCClient client = new GRPCClient(alarmSetting.getTargetHost(), alarmSetting.getTargetPort());
+            client.connect();
+            alarmServiceStub = AlarmServiceGrpc.newStub(client.getChannel());
+        }
+    }
+
+    @Override public void doAlarm(List<AlarmMessage> alarmMessage) {
+
+        if (alarmServiceStub == null) {
+            return;
+        }
+
+        StreamObserver<org.apache.skywalking.oap.server.core.alarm.grpc.AlarmMessage> streamObserver =
+            alarmServiceStub.withDeadlineAfter(10, TimeUnit.SECONDS).doAlarm(new StreamObserver<Response>() {
+                @Override public void onNext(Response response) {
+
+                }
+
+                @Override public void onError(Throwable throwable) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Send alarm message failed: {}", throwable.getMessage());
+                    }
+
+                }
+
+                @Override public void onCompleted() {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Send alarm message successful.");
+                    }
+                }
+            });
+
+        alarmMessage.forEach(message -> {
+            org.apache.skywalking.oap.server.core.alarm.grpc.AlarmMessage.Builder builder =
+                org.apache.skywalking.oap.server.core.alarm.grpc.AlarmMessage.newBuilder();
+
+            builder.setScopeId(message.getScopeId());
+            builder.setScope(message.getScope());
+            builder.setName(message.getName());
+            builder.setId0(message.getId0());
+            builder.setId1(message.getId1());
+            builder.setRuleName(message.getRuleName());
+            builder.setAlarmMessage(message.getAlarmMessage());
+            builder.setStartTime(message.getStartTime());
+
+            streamObserver.onNext(builder.build());
+        });
+
+        streamObserver.onCompleted();
+
+        if (log.isDebugEnabled()) {
```

Comment:
If you are using gRPC stream, you should use wait for the onComplete event to avoid the memory leak. What do you think?

Reply:
Yeah

Reply:
Waiting for your update.